### ISSUE TITLE
2.4.1 says Syntax error, but code runs fine

### ISSUE THREAD

**Issue by ElvenSpellmaker** [rodjek/puppet-lint#891]
Hey, dependabot has opened a PR To upgrade us from 2.3.6 to 2.4.1 which fails our build:

![image](https://user-images.githubusercontent.com/2286713/66470952-7882d900-ea82-11e9-9ed7-6e5009827e4f.png)

![image](https://user-images.githubusercontent.com/2286713/66470934-715bcb00-ea82-11e9-92f1-c6d3bfab234e.png)

However this file is working fine in our environment and I can't see a syntax error here.

**Comment by usev6** [comment]
Hi @ElvenSpellmaker, I'm not able to reproduce the failure with the code from your screenshot. (Assuming there are no weird/hidden UTF8 characters involved.)

```
$ cat gh/manifests/issue891.pp
# gh issue 891
class gh::issue891 {
  exec { $install_title:
    command => "cmd /C ${destination_directory}\\${destination_file} /quiet /norestart ${extra_msi_flags}",
    path    => $::path,
    require => Download_file[$download_file],
    creates => $creates_path,
  }
}
$ puppet-lint gh/manifests/issue891.pp
WARNING: top-scope variable being used without an explicit namespace on line 3
WARNING: top-scope variable being used without an explicit namespace on line 4
WARNING: top-scope variable being used without an explicit namespace on line 4
WARNING: top-scope variable being used without an explicit namespace on line 4
WARNING: top-scope variable being used without an explicit namespace on line 6
WARNING: top-scope variable being used without an explicit namespace on line 7
$ puppet-lint --version
puppet-lint 2.4.1
```

It doesn't talk about a syntax error, so the problem seems to be somewhere else. Could you please provide more context (ideally with the code being reduced to the problematic resource/whatever).

**Comment by ElvenSpellmaker** [comment]
Hey @usev6, below is the full file that triggers the warning, line 43 is where the syntax error is reported.

```puppet
define profile::applications::msi_install(
  $msi_tools,
  $download_path_suffix,
  $creates_path,
  $msi_flags = undef,
) {

  $temp_directory         = 'C:\\Temp'
  $destination_directory  = "${temp_directory}\\${download_path_suffix}"

  ensure_resource('File', $temp_directory,
    {
      ensure => directory,
    }
  )
  ensure_resource('File', $destination_directory,
    {
      ensure  => directory,
      require => File[$temp_directory],
    }
  )

  $extra_msi_flags = $msi_flags ? {
    undef   => '',
    default => $msi_flags,
  }

  $msi_tools.each |$tool_name, $url| {

    $download_file    = "${destination_directory}-${tool_name}"
    $install_title    = "${download_file}-install"
    $destination_file = "${tool_name}.msi"
    $creates_path     = "${creates_path}\\${tool_name}\\"

    download_file { $download_file:
      url                   => $url,
      destination_directory => $destination_directory,
      destination_file      => $destination_file,
      require               => File[$destination_directory],
    }

    exec { $install_title:
      command => "cmd /C ${destination_directory}\\${destination_file} /quiet /norestart ${extra_msi_flags}",
      path    => $::path,
      require => Download_file[$download_file],
      creates => $creates_path,
    }
  }
}
```

**Comment by usev6** [comment]
Thanks @ElvenSpellmaker, I'm able to reproduce the result now.

Looking for a minimal test case that still fails, I arrived at the following code (that should be valid). (It's possible that I removed to much code and arrived at a different problem, but I think, that's still the same problem.)
```
# gh issue 891
class gh::issue891 {
  $var1 = "${content1}\\"
  $var2 = "${content2}"
}
```

What confuses me a bit: My golfed code and your complete code both fail for me on version 2.3.6, too. If I understood you correctly, you only saw the error after an (attempted) update from 2.3.6 to 2.4.1. So, with 2.3.6 you didn't get an error with your complete code?

**Comment by ElvenSpellmaker** [comment]
Indeed, currently it works with 2.3.6 for me:
![image](https://user-images.githubusercontent.com/2286713/66574595-67ae9200-eb6c-11e9-98e6-6fbe0f861228.png)
![image](https://user-images.githubusercontent.com/2286713/66574611-6e3d0980-eb6c-11e9-8203-5e9982251a03.png)

**Comment by usev6** [comment]
Oh, I mixed something up. It doesn't die for me either with 2.3.6. (Both the full and the golfed version.)

**Comment by usev6** [comment]
The following patch seems to help and it looks correct to me (the quantifier seems to be wrong):
```
diff --git a/lib/puppet-lint/lexer/string_slurper.rb b/lib/puppet-lint/lexer/string_slurper.rb
index bd76a25..9556c2c 100644
--- a/lib/puppet-lint/lexer/string_slurper.rb
+++ b/lib/puppet-lint/lexer/string_slurper.rb
@@ -15,7 +15,7 @@ class PuppetLint
       END_INTERP_PATTERN = %r{\}}
       END_STRING_PATTERN = %r{(\A|[^\\])(\\\\)*"}
       UNENC_VAR_PATTERN = %r{(\A|[^\\])\$(::)?(\w+(-\w+)*::)*\w+(-\w+)*}
-      ESC_DQUOTE_PATTERN = %r{\\+"}
+      ESC_DQUOTE_PATTERN = %r{\\"}
       LBRACE_PATTERN = %r{\{}
 
       def initialize(string)
```

I'll think about a proper test and will open a PR.

**Comment by rodjek** [comment]
Fixed by #892, @cdenneen a new release will be out in a couple of hours.
